src/remote.c

```diff
@@ -670,21 +670,32 @@ int git_remote_set_pushurl(git_repository *repo, const char *remote, const char*
 	return set_url(repo, remote, CONFIG_PUSHURL_FMT, url);
 }
 
-const char* git_remote__urlfordirection(git_remote *remote, int direction)
+static int git_remote__urlresolve(const char *url, char **resolved_url, int direction, const git_remote_callbacks *callbacks)
 {
-	assert(remote);
+	if (callbacks && callbacks->resolve_url) {
+		return callbacks->resolve_url(url, resolved_url, direction, callbacks->payload);
+	}
+
+	*resolved_url = url ? git__strdup(url) : NULL;
+
+	return GIT_OK;
+}
 
+int git_remote__urlfordirection(struct git_remote *remote, int direction, const git_remote_callbacks *callbacks, char **url_out)
+{
+	assert(remote);
 	assert(direction == GIT_DIRECTION_FETCH || direction == GIT_DIRECTION_PUSH);
 
 	if (direction == GIT_DIRECTION_FETCH) {
-		return remote->url;
+		return git_remote__urlresolve(remote->url, url_out, direction, callbacks);
 	}
 
 	if (direction == GIT_DIRECTION_PUSH) {
-		return remote->pushurl ? remote->pushurl : remote->url;
+		return git_remote__urlresolve(remote->pushurl ? remote->pushurl : remote->url, url_out, direction, callbacks);
 	}
 
-	return NULL;
+	/* unreachable */
```

Comment:
Not exactly unreachable :P In release-code, the above assert would not exist and thus callers may reach this statement

Reply:
So just remove the comment and return an error?

Reply:
Yeah, guess that's fine.